phabricatoremails/mail.py

```diff
@@ -55,8 +59,39 @@ def encode_from(self, from_address):
             return from_address
 
     def to_mime_message(self, from_address, include_target_in_subject=False):
-        msg = MIMEMultipart("alternative")
+        # There's four email threading headers that we use:
+        # * Thread-Topic and Thread-Index are for Outlook.
+        # * In-Reply-To and References are RFC 2822 compliant and and used by
+        #   almost everything else.
+        # Note: just like Phabricator, we don't track the parent-child relationship of
+        # each message as a long interconnected chain. Instead, we configure each
+        # message such that it "References"/is "In-Reply-To" an imaginary first message
+        # (defined here as "thread_message_id").
+        # This is sufficient for email clients to associate emails together into a
+        # thread. Then, we have them sort the messages either by timestamp or the
+        # trailer of "Thread-Index" (for Outlook).
+        thread_id = f"revision-{self.revision_id}"
+        domain = from_address.split("@")[1]
+        thread_message_id = f"<{thread_id}@{domain}>"
+
+        # The Thread-Index header is used by Outlook to group and sort email threads.
+        # It's format is:
+        # * 27 bytes that are unique to the email thread.
+        # * a byte containing a space character
+        # * A 4-byte trailer used for ordering the messages.
+        # Our implementation here is heavily inspired by that of Phabricator's
+        # "generateThreadIndex()". We generate our unique first 27 bytes by
+        # hashing our unique thread id, then encode the current timestamp as
+        # the last 4 bytes.
+        thread_id_md5 = hashlib.md5(thread_id.encode("utf-8")).hexdigest()
+        thread_index_base = thread_id_md5.encode("utf-8")[:27]
+        thread_index = thread_index_base + b" " + struct.pack(">i", self.timestamp)
```

Comment:
This should probably all go into its own method, say `generate_thread_params` that takes `from_address` and possibly `revision_id` (unless you turn it into an instance method).  You could even return a dictionary with the required keys below and just update `msg` directly.